Cubical/Algebra/CommRing/Ideal.agda

```diff
@@ -70,10 +86,239 @@ module _ (R' : CommRing ℓ) where
+                              (sym (cong (α i ·_) βi≡∑γδ ∙∙ ∑Mulrdist (α i) (λ j → γ j · δ j)
+                                                         ∙∙ ∑Ext (λ j → ·Assoc (α i) (γ j) (δ j))))
+                              (∑Closed ((I ·i J) ·i K) (λ j → α i · γ j · δ j) -- each αᵢγⱼδⱼ∈IJK
+                                       λ j → prodInProd (I ·i J) K _ _
+                                               (prodInProd I J _ _ (α∈I i) (γ∈J j)) (δ∈K j)))
+                (β∈JK i))
+
+  incl2 : ((I ·i J) ·i K) .fst ⊆ (I ·i (J ·i K)) .fst
+  incl2 x = rec isPropPropTrunc
+          λ (_ , (α , β) , α∈IJ , β∈K , x≡∑αβ)
+              → subst-∈ ((I ·i (J ·i K)) .fst) (sym x≡∑αβ)
+                (∑Closed (I ·i (J ·i K)) (λ i → α i · β i)
+          λ i → rec isPropPropTrunc
+                (λ (_ , (γ , δ) , γ∈I , δ∈J , αi≡∑γδ)
+                   → subst-∈ ((I ·i (J ·i K)) .fst)
+                              (sym (cong (_· β i) αi≡∑γδ ∙∙ ∑Mulldist (β i) (λ j → γ j · δ j)
+                                         ∙∙ ∑Ext (λ j → sym (·Assoc (γ j) (δ j) (β i)))))
+                              (∑Closed (I ·i (J ·i K)) (λ j → γ j · (δ j · β i))
+                                       λ j → prodInProd I (J ·i K) _ _ (γ∈I j)
+                                               (prodInProd J K _ _ (δ∈J j) (β∈K i))))
+                (α∈IJ i))
+
+ ·iRdist+i : ∀ (I J K : CommIdeal) → I ·i (J +i K) ≡ I ·i J +i I ·i K
+ ·iRdist+i I J K = CommIdeal≡Char incl1 incl2
+  where
+  incl1 : (I ·i (J +i K)) .fst ⊆ (I ·i J +i I ·i K) .fst
+  incl1 x = rec isPropPropTrunc
+    λ (n , (α , β) , α∈I , β∈J+K , x≡∑αβ) → subst-∈ (((I ·i J) +i (I ·i K)) .fst) (sym x≡∑αβ)
+      (∑Closed ((I ·i J) +i (I ·i K)) (λ i → α i · β i) -- each αi·βi ∈ IJ+IK
+      λ i → rec isPropPropTrunc
+            (λ ((γi , δi) , γi∈J , δi∈K , βi≡γi+δi) →
+               ∣ (α i · γi , α i · δi) , prodInProd I J _ _ (α∈I i) γi∈J
+                                       , prodInProd I K _ _ (α∈I i) δi∈K
+                                       , cong (α i ·_) βi≡γi+δi ∙ ·Rdist+ _ _ _ ∣)
+            (β∈J+K i))
+
+  incl2 : ((I ·i J) +i (I ·i K)) .fst ⊆ (I ·i (J +i K)) .fst
+  incl2 x = rec isPropPropTrunc λ ((y , z) , y∈IJ , z∈IK , x≡y+z)
+          → subst-∈ ((I ·i (J +i K)) .fst) (sym x≡y+z)
+              ((I ·i (J +i K)) .snd .+Closed (inclHelperLeft _ y∈IJ) (inclHelperRight _ z∈IK))
+   where
+   inclHelperLeft : (I ·i J) .fst ⊆ (I ·i (J +i K)) .fst
+   inclHelperLeft x' = map (λ (n , (α , β) , α∈I , β∈J , x'≡∑αβ)
+                     → n , (α , β) , α∈I , (λ i → +iLincl J K _ (β∈J i)) , x'≡∑αβ)
+
+   inclHelperRight : (I ·i K) .fst ⊆ (I ·i (J +i K)) .fst
+   inclHelperRight x' = map (λ (n , (α , β) , α∈I , β∈K , x'≡∑αβ)
+                      → n , (α , β) , α∈I , (λ i → +iRincl J K _ (β∈K i)) , x'≡∑αβ)
+
+ -- only one absorption law, i.e. CommIdeal , +i , ·i does not form a dist. lattice
+ ·iAbsorb+i : ∀ (I J : CommIdeal) → I +i (I ·i J) ≡ I
```

Comment:
Two general remark: 
- Maybe it would be useful to state the various `⊆` lemmas on the top-level? Maybe some statements could even be generalized and the proofs be streamlined that way (e.g. `inclHelperLeft` and `inclHelperRight` just above looks very similar...)
- Maybe you should have a special version of `⊆` for ideals to avoid having to put `fst` on both sides all the time?

Reply:
I'm not too happy with the way things are now either. I think moving up all the interesting inclusions is a good first step. I did start working on replacing the usual path types with the equivalent double inclusion:
https://github.com/mzeuner/cubical/commit/4652ee9a9e12e201e07d36fe372917e88006e735

Maybe it's worthwhile doing that afterall. Adding inclusion and containment of ideals is a good idea to make things more readable. The only issue I'm somewhat unhappy about is that you want to have some sort `cong` for subsets and ideals like this
https://github.com/mzeuner/cubical/blob/Zariski_Lattice/Cubical/Foundations/Powerset.agda#L82

It feels a bit like cheating because you go up a universe level and back down but maybe I'm just being fussy here...

Reply:
I'm a bit puzzled why you might need these lemmas? You can surely just put the results you have on the top-level without using any lemmas like this. Anyway I'm not sure how useful `≡ᴾ` is and it might be best to just use `≡` when possible but state the containment lemmas separately so that you can reuse them

Reply:
Changed everything according to your suggestions.